src/video.tsx

```diff
@@ -56,27 +104,57 @@ const Video = React.forwardRef<HTMLVideoElement, VideoProps>(
             muted = true,
             loop = true,
             preload = 'none',
+            loadingTimeout = 10000,
+            onLoad,
+            onError,
+            onPause,
+            onPlay,
             aspectRatio = '1:1',
             dataAttributes,
             ...props
         },
         ref
     ) => {
-        const borderRadiusContext = useMediaBorderRadius();
+        const [videoStatus, dispatch] = React.useReducer(videoReducer, 'loading');
+        const videoRef = React.useRef<HTMLVideoElement | null>(null);
+        const loadedSource = React.useRef<VideoSource>();
 
+        const borderRadiusContext = useMediaBorderRadius();
         const ratio = typeof aspectRatio === 'number' ? aspectRatio : RATIO[aspectRatio];
 
-        const videoRef = React.useRef<HTMLVideoElement | null>(null);
+        const handleError = React.useCallback(() => {
+            if (videoStatus === 'loading') {
+                dispatch('fail');
+                onError?.();
+            }
+        }, [onError, videoStatus]);
 
         React.useEffect(() => {
+            if (loadedSource.current !== src) {
+                loadedSource.current = src;
+                const loadingTimeoutId = setTimeout(handleError, loadingTimeout);
+                dispatch('reset');
+                videoRef.current?.load();
+
+                return () => {
+                    clearTimeout(loadingTimeoutId);
+                };
+            }
+        }, [src, loadingTimeout, handleError]);
+
+        const onLoadFinish = () => {
+            onLoad?.();
             const video = videoRef.current;
-            if (video && autoPlay && video.paused) {
+            const shouldAutoPlay = autoPlay && !isRunningAcceptanceTest();
```

Comment:
I feel like this `isRunningAcceptanceTest` logic is repeated. Maybe the default value to false isn't needed if you are already checking it here